app/src/main/cpp/skyline/input/npad_device.cpp

```diff
@@ -0,0 +1,293 @@
+                section.header.singleColor = {0, 0};
+                break;
+
+            case NpadControllerType::Handheld:
+            case NpadControllerType::JoyconDual:
+                section.header.singleColorStatus = NpadColorReadStatus::Disconnected;
+                section.header.dualColorStatus = NpadColorReadStatus::Success;
+                section.header.leftColor = {0, 0};
+                section.header.rightColor = {0, 0};
+                break;
+
+            case NpadControllerType::None:
+                break;
+        }
+
+        section.singleBatteryLevel = NpadBatteryLevel::Full;
+        section.leftBatteryLevel = NpadBatteryLevel::Full;
+        section.rightBatteryLevel = NpadBatteryLevel::Full;
+
+        type = newType;
+        controllerInfo = &GetControllerInfo();
+
+        updateEvent->Signal();
+    }
+
+    void NpadDevice::Disconnect() {
+        if (type == NpadControllerType::None)
+            return;
+
+        section = {};
+        explicitAssignment = false;
+        globalTimestamp = 0;
+
+        type = NpadControllerType::None;
+        controllerInfo = nullptr;
+
+        updateEvent->Signal();
+    }
+
+    NpadControllerInfo &NpadDevice::GetControllerInfo() {
+        switch (type) {
+            case NpadControllerType::ProController:
+                return section.fullKeyController;
+            case NpadControllerType::Handheld:
+                return section.handheldController;
+            case NpadControllerType::JoyconDual:
+                return section.dualController;
+            case NpadControllerType::JoyconLeft:
+                return section.leftController;
+            case NpadControllerType::JoyconRight:
+                return section.rightController;
+            default:
+                throw exception("Cannot find corresponding section for ControllerType: {}", type);
+        }
+    }
+
+    NpadControllerState &NpadDevice::GetNextEntry(NpadControllerInfo &info) {
+        auto &lastEntry = info.state.at(info.header.currentEntry);
+
+        info.header.timestamp = util::GetTimeTicks();
+        info.header.entryCount = std::min(static_cast<u8>(info.header.entryCount + 1), constant::HidEntryCount);
+        info.header.currentEntry = (info.header.currentEntry != constant::HidEntryCount - 1) ? info.header.currentEntry + 1 : 0;
+
+        auto &entry = info.state.at(info.header.currentEntry);
+
+        entry.globalTimestamp = globalTimestamp;
+        entry.localTimestamp = lastEntry.localTimestamp + 1;
+        entry.buttons = lastEntry.buttons;
+        entry.leftX = lastEntry.leftX;
+        entry.leftY = lastEntry.leftY;
+        entry.rightX = lastEntry.rightX;
+        entry.rightY = lastEntry.rightY;
+        entry.status.raw = connectionState.raw;
+
+        return entry;
+    }
+
+    void NpadDevice::SetButtonState(NpadButton mask, NpadButtonState state) {
+        if (!connectionState.connected || !controllerInfo)
```

Comment:
`connectionState.connected` isn't set in handheld, maybe have a private bool with the connection state that's set at the end of `Connect` that way you can drop the other check.